proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,281 @@
+# Overview and goals
+
+Apri 20th, 2020.
+
+This document provides options and recommendations for 5.0 to handle overlapping serializer goals including reach, faster cold startup performance, reducing memory usage, and smaller size-on-disk.
+
+## Reach
+Achieved by limiting or removing dependencies on Reflection.Emit and `Type.MakeGenericType`. This allows the serializer to be used on additional runtimes including [Xamarin iOS](https://github.com/dotnet/runtime/issues/31326).
+
+Ideally we add a high-performance emit replacement for targeted scenarios such as property setters, getters and constructors.
+
+Some serializers do not have a non-emit fallback. Although System.Text.Json has a fallback for emit, it does not have a fallback for its usage of `Type.MakeGenericType`.
+
+## Faster startup performance
+
+This is important for scenarios including [micro-services](https://github.com/dotnet/runtime/issues/1568) where processes start and exit frequently. It is also important for benchmarks including TechEmpower.
+
+The serializer has fast startup compared to most other serializers including Newtonsoft's JSON.Net, JIL and Utf8Json. Startup perf is mostly affected by generating IL - the Jil serializer is the most aggressive here and has the highest startup cost.
+
+There is no stated goal for 5.0 to make the steady-state CPU performance faster than it is now in 5.0 for general sceanrios, although we do not want to noticably regress steady-state performance while addressing the stated goals including improving startup performance. It is possible some code-gen options described later may have steady-state performance gains, but again that is not a stated goal.
+
+<details>
+  <summary>5.0 master startup benchmarks</summary>
+
+A small POCO was used (`LoginViewModel` benchmark test class).
+
+| Serializer |  Serialize (us)  | Serialize ratio | Deserialize (us) | Deserialize ratio |
+| :-- | :-- | :-- | :-- | :--
+| **System.Text.Json** | 4,720 | 1.00 | 19,379 | 1.00 
+| **Json.NET** | 24,916 | 5.28 | 102,226 | 5.28 
```

Comment:
What's the ReadyToRun setting for these measurements? If things were left at their defaults, System.Text.Json will be compiled with ReadyToRun because that's how we ship the library out of CoreFX and none of the other libraries will get that benefit. ReadyToRun has huge impact on startup time.

Reply:
cc @layomia 

For the "overhead" status I provided later (e.g. 22ms overhead) those were done with a public 5.0 build.

Reply:
The ReadyToRun setting for these benchmarks is `false`.

I ran them again with `ReadyToRun` set to true using a Release build of .NET 5 at https://github.com/dotnet/runtime/commit/38c2d5513c - https://github.com/layomia/jsonconvertergenerator/blob/00e81d5411175c8874528f2837bfbf3319637426/run_benchmarks.py#L31-L32

Deserialize LoginViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| System.Text.Json   | 4351      | 1.00  |
| Json.NET           | 2417      | 0.56  |
| Utf8Json           | 732       | 0.17  |
| Jil                | 45095     | 10.36 |

Serialize LoginViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| System.Text.Json   | 20885     | 1.00  |
| Json.NET           | 82531     | 3.95  |
| Utf8Json           | 71666     | 3.43  |
| Jil                | 72114     | 3.45  |

<details>
<summary>
See more benchmarks for other POCOs. The trend is we have faster start-up perf than the other serializers when serializing, but  we are slower when deserializing (except for Jil).
</summary>

---

`Default` = System.Text.Json

Deserialize LoginViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 45095     | 10.36 |
| Json.NET           | 2417      | 0.56  |
| Utf8Json           | 732       | 0.17  |
| Default            | 4351      | 1.00  |

Deserialize Location

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 55325     | 10.52 |
| Json.NET           | 3304      | 0.63  |
| Utf8Json           | 1377      | 0.26  |
| Default            | 5257      | 1.00  |

Deserialize IndexViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 80323     | 10.49 |
| Json.NET           | 5327      | 0.70  |
| Utf8Json           | 2867      | 0.37  |
| Default            | 7654      | 1.00  |

Deserialize MyEventsListerViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 84313     | 10.33 |
| Json.NET           | 5229      | 0.64  |
| Utf8Json           | 3554      | 0.44  |
| Default            | 8161      | 1.00  |

Deserialize CollectionsOfPrimitives

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 49745     | 3.80  |
| Json.NET           | 5249      | 0.40  |
| Utf8Json           | 2545      | 0.19  |
| Default            | 13087     | 1.00  |

Serialize LoginViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 72114     | 3.45  |
| Json.NET           | 82531     | 3.95  |
| Utf8Json           | 71666     | 3.43  |
| Default            | 20885     | 1.00  |

Serialize Location

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 71383     | 3.28  |
| Json.NET           | 84474     | 3.88  |
| Utf8Json           | 70136     | 3.23  |
| Default            | 21744     | 1.00  |

Serialize IndexViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 92385     | 3.54  |
| Json.NET           | 89281     | 3.42  |
| Utf8Json           | 81163     | 3.11  |
| Default            | 26090     | 1.00  |

Serialize MyEventsListerViewModel

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 116070    | 3.76  |
| Json.NET           | 91335     | 2.96  |
| Utf8Json           | 89143     | 2.89  |
| Default            | 30833     | 1.00  |

Serialize CollectionsOfPrimitives

| Test               | Mean (us) | Ratio |
|--------------------|-----------|-------|
| Jil                | 74404     | 2.53  |
| Json.NET           | 85696     | 2.92  |
| Utf8Json           | 76411     | 2.60  |
| Default            | 29367     | 1.00  |

</details>